### PR TITLE
Cache extracted EmbeddedXlTables based on xlsx file hashes

### PR DESCRIPTION
**Pull request by siddharth-krishna**
This PR adds a feature to cache the extracted EmbeddedXlTables from input XLSX files by default (and removes the `--use_pkl` flag). This is because on some bigger benchmarks `openpyxl` takes ages to extract the tables, and the xl2times developer workflow rarely involves changing the XLSX files.

The cache is stored in `xl2times/.cache/` and saves files based on the hash of the XLSX file. It also checks filename (and can be extended to check modification time) to avoid hash collisions. Another future improvement can be to have a maximum size for the cache directory.

~Change in runtime appears to be about 50% on CI (note that for this PR's CI, the cache is created when the tool is run first on the PR branch, and then the cache is used when it is run on `main`):~
(See comment below for more accurate performance implications)
image

There's also a new flat `--no_cache` to ignore the cache directory and read the files from scratch.

@SamRWest just tagging you so you are aware of this (potentially confusing) behaviour change.

### CODE REVIEW

**Comment by siddharth-krishna** [comment]
CI is failing because there is no cache present yet, and regression tests run first on the PR branch and second on the main branch -- the main branch uses the cache created by the PR branch, so is much faster!

When I get time, I'm planning to disable time-based regressions as a CI failure, and instead have CI comment on the PR with the difference in time if it is significant (since time measurements are anyway not super accurate as we run benchmarks in parallel). That way we can see the difference in time and decide whether to ignore / investigate as appropriate.

**Comment by olejandro** [comment]
Cool, thanks! Do I just merge it then?

**Comment by siddharth-krishna** [comment]
Not yet, I need to fix the cache in the GitHub Actions

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -14,11 +14,24 @@ jobs:
...
     runs-on: ubuntu-latest
 
+    env:
+      PY_VERSION: "3.11"
+      REF_TIMES_model: "b488fb07f0899ee8b7e710c230b1a9414fa06f7d"
+      REF_demos-xlsx: "f956db07a253d4f5c60e108791ab7bb2b8136690"
+      REF_demos-dd: "2848a8a8e2fdcf0cdf7f83eefbdd563b0bb74e86"
+      REF_tim: "e820d8002adc6b1526a3bffcc439219b28d0eed5"
+      REF_tim-gams: "703f6a4e1d0bedd95c3ebdae534496f3a7e1b7cc"
+
```
Comment (olejandro) [inline .github/workflows/ci.yml:24]: Here we are fixing the versions to be used for tests on the CI, right?
Great!

**Review by olejandro — APPROVED** [review]

**Comment by siddharth-krishna** [comment]
It looks like I was mistaken about the performance implications. Here are the total runtimes of running all the benchmarks on my laptop:

```
main         812s
PR-first-run 871s
PR-secd-run  658s
```

So the act of hashing and creating the cache in the first run adds 59s (7%) runtime, but once we have the cache, subsequent runs are 152s (19%) faster. I'm going to merge this in, assuming that that tradeoff is useful for now, but happy to discuss whether we want caching to be the default behaviour and iterate on this further.

**Comment by olejandro** [comment]
Thanks @siddharth-krishna! I believe this will definitely be useful, also when using the tool  for scenario analysis. I guess, later on, we could develop this further to skip some of the transforms depending on what's changed?

@SamRWest how long does it take to read AusTIMES?

**Comment by olejandro** [comment]
In the case of TIMES-US Model (TUSM), reading of the files takes about 6 minutes (ca. 25% of the total processing time). Although `process_wildcards` takes most of the time, i.e. 63%, caching gives the overall processing time a nice boost!

**Comment by SamRWest** [comment]
Nice - you beat me to it! I was considering adding basically this feature to speed up my AusTIMES runs :) Thanks!

It might be worth trying caching with parquet files instead of pickles in future, they're usually lightning fast to read/write compared to pickles, and the file sizes are much smaller. The downside is that you'd have to save one file per table though.

Speeds for Austimes are:
With caching, first run:
`Extracted (potentially cached) 555 tables, 34360 rows in 0:01:32.051419`
With caching, second run:
`Extracted (potentially cached) 555 tables, 34360 rows in 0:00:00.842077`
Nice speedup :)

**Comment by siddharth-krishna** [comment]
Great to hear this is useful! Thanks for the suggestions, I've moved the discussion to an issue so we don't lose track of it: #199

**Pull request merged**